run/fvde2john.py

```diff
@@ -0,0 +1,220 @@
+
+def uint_to_int(b):
+    return int(b[::-1].hex(), 16)
+
+def guid_to_hex(guid):
+    guid_parts = guid.split('-')
+
+    hex_str = ''.join([guid_parts[0][i:i+2] for i in range(0, len(guid_parts[0]), 2)][::-1])
+    hex_str += ''.join([guid_parts[1][i:i+2] for i in range(0, len(guid_parts[1]), 2)][::-1])
+    hex_str += ''.join([guid_parts[2][i:i+2] for i in range(0, len(guid_parts[2]), 2)][::-1])
+    hex_str += guid_parts[3]
+    hex_str += guid_parts[4]
+
+    return hex_str.lower()
+
+def parse_partition_table(fp):
+    # determine whether sector size is 0x200 or 0x1000
+    sector_size = 0x0
+
+    # look for EFI PART at start of sector 1
+    fp.seek(0x200)
+    signature = fp.read(0x8)
+    if signature == b'EFI PART':
+        sector_size = 0x200
+
+    else:
+        fp.seek(0x1000)
+        signature = fp.read(0x8)
+        if signature == b'EFI PART':
+            sector_size = 0x1000
+
+    print("[+] Identified sector size:", sector_size)
+
+    if not sector_size:
+        print(f"[!] Invalid sector size {sector_size} (not 512 or 4096 bytes). Exiting.")
+
+    fp.seek(2 * sector_size) # go to sector 2
+    partitions = []
+    partition_entry = b'1'
+    while any(partition_entry):
+        partition_entry = fp.read(128)
+        if any(partition_entry):
+            partitions.append(partition_entry)
+
+    partition_dict = {}
+    for p in partitions:
+        part_GUID, type_GUID, start, partition_name = parse_partition_entry(p)
+        sp = uint_to_int(start) * sector_size
+        partition_dict[part_GUID.hex()] = {'start':sp, 'partition_type':type_GUID.hex(), 'partition_name':partition_name.decode('utf-16').strip('\x00')}
+
+    return partition_dict
+
+def findall(p, s):
+    i = s.find(p)
+    while i != -1:
+        yield i
+        i = s.find(p, i+1)
+
+def parse_partition_entry(partition_entry):
+    type_GUID = partition_entry[0:0x10]
+    part_GUID = partition_entry[0x10:0x20]
+    start_LBA = partition_entry[0x20:0x28]
+    partition_name = partition_entry[0x38:0x80]
+    return part_GUID, type_GUID, start_LBA, partition_name
+
+def parse_corestorage_header(fp, start_pos):
+    fp.seek(start_pos + 176)
+    aes_key = fp.read(0x10)
+    return aes_key
+
+def AES_XTS_decrypt(aes_key, tweak, ct):
+    decryptor = Cipher(
+        algorithms.AES(key=aes_key + b'\x00' * 16),
+        modes.XTS(tweak=tweak),
+    ).decryptor()
+    pt = decryptor.update(ct)
+    return pt
+
+def parse_keybag_entry(uuid, pt):
+    uuid_iterator = findall(uuid, pt)
+    for sp in uuid_iterator:
+        ke_uuid, ke_tag, ke_keylen = pt[sp:sp+16], uint_to_int(pt[sp + 16:sp + 18]), uint_to_int(pt[sp + 18:sp + 20])
+        padding = pt[sp + 20:sp + 24]
+        keydata = pt[sp + 24: sp + 24 + ke_keylen]
+
+        # only tag 3 is needed for constructing the hash
+        if ke_tag == 3:
+            assert padding == b'\x00\x00\x00\x00'
+            volume_unlock_record = keydata
+
+            return volume_unlock_record
```

Comment:
This can be simplified.

```suggestion
            return keydata
```

Reply:
I didn't check, but maybe this redundancy adds an oz of self documentation in line with some spec?

Reply:
Yes this was from following the documentation, but will simplify anyway